(Feature) Decode multiSend tx details

fernandomg opened the pull request:
This PR closes #865, by properly presenting tx's decoded data for `multiSend` txs in the tx's expanded view.

<img alt="multiSend-details-v2" src="https://user-images.githubusercontent.com/3315606/86968107-f96e1e80-c141-11ea-9361-d193154ce875.gif" width="550" />



~**note:** I converted it to draft because, despite it's functional and can be tested, there are pending components refactor required. See [42ed908](https://github.com/gnosis/safe-react/pull/1106/commits/42ed90823d2b454ad62902aecbce8e930ff3f5a8)~

gnosis-info commented:
Travis automatic deployment:
 https://pr1106--safereact.review.gnosisdev.com/app

gnosis-info commented:
Travis automatic deployment:
 https://pr1106--safereact.review.gnosisdev.com/app

gnosis-info commented:
Travis automatic deployment:
 https://pr1106--safereact.review.gnosisdev.com/app

gnosis-info commented:
Travis automatic deployment:
 https://pr1106--safereact.review.gnosisdev.com/app

lukasschor commented:
I loaded this Safe to see how a Sablier Safe App interaction looks like with decoded Multisend: 0xCfF7eB7C0cAc85188dF7748486A3ab8a20F56315

While doing that, I saw that there is a "Cancellation transaction" being displayed, which is not the case on the production version: 
![image](https://user-images.githubusercontent.com/17196842/87314757-fa2cf900-c523-11ea-8a1f-d0671066c71b.png)

So seems like there is something broken with the tx <> cancellation transaction merging.

gnosis-info commented:
Travis automatic deployment:
 https://pr1106--safereact.review.gnosisdev.com/app

lukasschor commented:
![image](https://user-images.githubusercontent.com/17196842/87416545-26518400-c5cf-11ea-93f6-5d610ce68658.png)

Could we make the function & parameter names bold?

Also, Richard had the idea of using the function name in the overview to make it easier to see which Transaction should be expanded to get some information on specific things. So instead of "Transaction 1" it could say "Transaction 1 (approve)". I think this might be an improvement that does not require too much effort, but happy to get your thoughts on this. I'm also fine leaving it out and opening a separate ticket for this.

francovenica commented:
Regarding the names of the Tx's. They are shown in the Tx details but not in the app operation itself:
Example: this is a "Supply" in compound, but in the tx details the only thing you can see is the "mint" and "approve" so you don't know that it was a Supply unless you remember it. Also the when you Supply the modal won't show that you are minting and approving
![image.png](https://images.zenhubusercontent.com/5e14b8e3edee21d6f9aa6c8a/3b810519-f93d-43c6-b5b2-7b6df0512d99)
![image.png](https://images.zenhubusercontent.com/5e14b8e3edee21d6f9aa6c8a/db8650fe-f853-443b-9fef-1ff6233ce000)

So we could put that this is a "Supply" operation and not just the "mint" and "approve" that is internally done. 

This is an example of one app. I still gotta see how it looks on the others

lukasschor commented:
> Regarding the names of the Tx's. They are shown in the Tx details but not in the app operation itself:
> Example: this is a "Supply" in compound, but in the tx details the only thing you can see is the "mint" and "approve" so you don't know that it was a Supply unless you remember it. Also the when you Supply the modal won't show that you are minting and approving
> ![image.png](https://camo.githubusercontent.com/1ee0c8649215e0c068ecb34151a7987369a083a5/68747470733a2f2f696d616765732e7a656e68756275736572636f6e74656e742e636f6d2f3565313462386533656465653231643666396161366338612f33623831303531392d663933642d343363362d623562322d376236646630353132643939)
> ![image.png](https://camo.githubusercontent.com/a74f3f1bcb70ec6e0cc5eb812bee78cdeb28d2be/68747470733a2f2f696d616765732e7a656e68756275736572636f6e74656e742e636f6d2f3565313462386533656465653231643666396161366338612f64623836353066652d663835332d343433622d396665662d316666363233336365303030)
> 
> So we could put that this is a "Supply" operation and not just the "mint" and "approve" that is internally done.
> 
> This is an example of one app. I still gotta see how it looks on the others

The issue is that the "Supply" name is a human-readable name given by the Safe App. But on the transaction level this information is basically encoded as a `approve` and `mint`. We cannot take any information from the Safe app to rely on, as this would create false security to the user and a malicious app could take advantage of this. So the base is the direct transaction data, which we then should aim to decode as much as possible over time to show to the user the consequences of the transaction in a human-readable manner.

gnosis-info commented:
Travis automatic deployment:
 https://pr1106--safereact.review.gnosisdev.com/app

gnosis-info commented:
Travis automatic deployment:
 https://pr1106--safereact.review.gnosisdev.com/app

lukasschor commented:
![image](https://user-images.githubusercontent.com/17196842/87718619-6755be00-c7b2-11ea-9c34-5035dec18a11.png)

Seems there is some whitespace missing. Only happens for this boolean parameter. In other cases it looks like this:

![image](https://user-images.githubusercontent.com/17196842/87718677-85232300-c7b2-11ea-9004-d39d45a5dc81.png)

gnosis-info commented:
Travis automatic deployment:
 https://pr1106--safereact.review.gnosisdev.com/app

fernandomg commented:
- changed label from `Transaction` to `Interaction`
- tx data displayed as stated in #1113 
- fixed details alignment styles
- fixed several layout issues (styles most)

**note:** merged with #934 as there are several typing fixes that prevented me from continuing here. So it's desirable not to merge this branch until #934 is merged.

gnosis-info commented:
Travis automatic deployment:
 https://pr1106--safereact.review.gnosisdev.com/app

francovenica commented:
I've checked a few tx created with the apps, the ones I can make in rinkeby. I'll check the apps that only work in mainnet once we deploy in stg mainnet.

It looks good to me

https://safe-team.dev.gnosisdev.com/app/#/safes/0x9913B9180C20C6b0F21B6480c84422F6ebc4B808/balances

gnosis-info commented:
Travis automatic deployment:
 https://pr1106--safereact.review.gnosisdev.com/app

gnosis-info commented:
Travis automatic deployment:
 https://pr1106--safereact.review.gnosisdev.com/app

gnosis-info commented:
Travis automatic deployment:
 https://pr1106--safereact.review.gnosisdev.com/app

gnosis-info commented:
Travis automatic deployment:
 https://pr1106--safereact.review.gnosisdev.com/app

gnosis-info commented:
Travis automatic deployment:
 https://pr1106--safereact.review.gnosisdev.com/app

fernandomg commented:
made a last change to support array of arguments to be displayed

this is a hacked view to see how data will be presented:
<img src="https://user-images.githubusercontent.com/3315606/87992281-4b0d9600-cabe-11ea-94f6-85b9c594356b.png" width="600" />

gnosis-info commented:
Travis automatic deployment:
 https://pr1106--safereact.review.gnosisdev.com/app

Agupane commented on src/components/Collapse/index.tsx:
```diff
@@ -7,14 +7,29 @@ import styled from 'styled-components'
+  title,
+  collapseClassName,
+  headerWrapperClassName,
+}) => {
```
Please add types

fernandomg replied:
:scream: completely missed those, thanks!

Agupane commented on src/logic/tokens/store/actions/fetchSafeTokens.ts:
```diff
@@ -14,8 +13,8 @@ import { SAFE_REDUCER_ID } from 'src/routes/safe/store/reducer/safe'
 
-const humanReadableBalance = (balance, decimals) => new BigNumber(balance).times(`1e-${decimals}`).toFixed()
 const noFunc = () => {}
 const updateSafeValue = (address) => (valueToUpdate) => updateSafe({ address, ...valueToUpdate })
```
Types

gnosis-info commented:
Travis automatic deployment:
 https://pr1106--safereact.review.gnosisdev.com/app

gnosis-info commented:
Travis automatic deployment:
 https://pr1106--safereact.review.gnosisdev.com/app

mikheevm commented:
@fernandomg you're the MVP :) you can ignore address book types, I'm working on it in contract types pr

fernandomg commented:
> @fernandomg you're the MVP :) you can ignore address book types, I'm working on it in contract types pr

thanks! I was going down the rabbit hole once again :smile: 

Will mark this PR as ready for review then

gnosis-info commented:
Travis automatic deployment:
 https://pr1106--safereact.review.gnosisdev.com/app

gnosis-info commented:
Travis automatic deployment:
 https://pr1106--safereact.review.gnosisdev.com/app

gnosis-info commented:
Travis automatic deployment:
 https://pr1106--safereact.review.gnosisdev.com/app

lukasschor commented:
![image](https://user-images.githubusercontent.com/17196842/88288320-e3d92700-ccf3-11ea-970d-9632ced9ffed.png)

Type shows unknown for Safe App interactions.

Safe: 0xCfF7eB7C0cAc85188dF7748486A3ab8a20F56315

fernandomg commented:
> <img src="https://user-images.githubusercontent.com/17196842/88288320-e3d92700-ccf3-11ea-970d-9632ced9ffed.png" width="300" />
> 
> Type shows unknown for Safe App interactions.
> 
> Safe: 0xCfF7eB7C0cAc85188dF7748486A3ab8a20F56315

@lukasschor, this issue doesn't seems to be related to this PR. It looks the same in: https://safe-team.dev.gnosisdev.com/app/#/safes/0xCfF7eB7C0cAc85188dF7748486A3ab8a20F56315/transactions
![image](https://user-images.githubusercontent.com/3315606/88289443-9cd73b80-cccb-11ea-8430-d7f65cc1f39f.png)

mikheevm commented:
@lukasschor @fernandomg 
this safe has invalid origin in transaction object ('1', 'compound')

an example of safe with correct one: https://safe-transaction.staging.gnosisdev.com/api/v1/safes/0xc3214e72A53695faC3A124C92FcA3f161D193299/transactions/?has_confirmations=True
this safe: https://safe-transaction.staging.gnosisdev.com/api/v1/safes/0xCfF7eB7C0cAc85188dF7748486A3ab8a20F56315/transactions/?has_confirmations=True

i'm not sure about a reason for this

francovenica commented:
I tried a few tx's today and they showed just fine.
![image.png](https://images.zenhubusercontent.com/5e14b8e3edee21d6f9aa6c8a/aae5246e-7be2-4199-9897-8ae4f68b7172)

I'd need to find a contract in rinkeby that deals with arrays in a realistic way

gnosis-info commented:
Travis automatic deployment:
 https://pr1106--safereact.review.gnosisdev.com/app

mikheevm commented on src/logic/tokens/store/actions/fetchSafeTokens.ts:
```diff
@@ -1,67 +1,92 @@
+    const currencyValues = currencyValuesSelector(state)
+
+    const { balances, currencyList, ethBalance, tokens }: ExtractedData = result.data.reduce(
+      extractDataFromResult(currentTokens),
```
https://stackoverflow.com/questions/28510753/when-should-i-use-withmutations-on-a-map-in-immutable-js

fernandomg replied:
Ok, once again. This change was just moving code around for better readability. We should address this in a different issue

mikheevm replied:
could you please create one?

fernandomg replied:
#1184

mikheevm commented on src/logic/tokens/store/actions/fetchSafeTokens.ts:
```diff
@@ -80,10 +105,7 @@ const fetchSafeTokens = (safeAddress: string) => async (
     const updateActiveTokens = activeTokens.equals(alreadyActiveTokens) ? noFunc : update({ activeTokens })
     const updateBalances = balances.equals(safeBalances) ? noFunc : update({ balances })
     const updateEthBalance = ethBalance === currentEthBalance ? noFunc : update({ ethBalance })
```
updating safe's properties in batch like this makes no sense because we're performing updates on the same instance. This is like sending 10 PUT requests to modify 10 properties on the same resource instead of sending one request

mikheevm replied:
Also I'm not sure if `.equals` performs checks for deep equality

fernandomg replied:
Ok, but we should aim to fix this in a different issue. All the changes in this file were related to typing fixes.

mikheevm replied:
Could you please create one?

fernandomg replied:
#1185

mikheevm commented on src/routes/safe/components/Transactions/TxsTable/ExpandedTx/TxDescription/CustomDescription.tsx:
```diff
@@ -0,0 +1,183 @@
+
+const MultiSigCustomData = ({ tx, order }: { tx: MultiSendDetails; order: number }): React.ReactElement => {
+  const classes = useStyles()
+  const methodName = tx.data?.method ? ` (${tx.data.method})` : ''
```
```suggestion
  const methodName = tx.data?.method?.toString() || ''
```

fernandomg replied:
but you're missing the parentheses that surround the method name

mikheevm replied:
oh, missed it, nevermind then

mikheevm commented on src/routes/safe/components/Transactions/TxsTable/ExpandedTx/TxDescription/CustomDescription.tsx:
```diff
@@ -0,0 +1,183 @@
+import React from 'react'
+import styled from 'styled-components'
+
+import RenderValue from './RenderValues'
```
Can we just name it value? It's quite obvious that the component renders something and we don't need to explicitly add it to the name

mikheevm commented on src/routes/safe/components/Transactions/TxsTable/ExpandedTx/TxDescription/CustomDescription.tsx:
```diff
@@ -0,0 +1,183 @@
+          {showTxData ? (
+            <>
+              {data}{' '}
+              <LinkWithRef
```
can we move the expand button somewhere else?  in case if expanded data is super long, user would have to scroll down for "show less" button

fernandomg replied:
but what if the user scrolled down to the end of the expanded data?

Anyway, this should be discussed in a different issue and with some discussion about UX

mikheevm commented on src/routes/safe/components/Transactions/TxsTable/columns.tsx:
```diff
@@ -56,8 +68,19 @@ export const getTxAmount = (tx, formatted = true) => {
+}
+
+const getIncomingTxTableData = (tx: Transaction): TableData => ({
+  [TX_TABLE_ID]: tx.blockNumber?.toString() ?? '',
```
what does happen to the table if there are multiple transactions with ids equal to ''?

fernandomg replied:
just :boom:

In fact, will happen what happens to rows with the same id. If you expand one, all the others will expand as well.

I thought about that when working on this and thought that the `blockNumber` isn't a proper id either, as it can be duplicated too.

fernandomg replied:
Ok, this is odd... it seems that the `id` is replaced by the function [`getTxTableData`](https://github.com/gnosis/safe-react/blob/b4ce1ff679fc31c2607a074f28bf73bba777ad97/src/routes/safe/components/Transactions/TxsTable/index.tsx#L38-L63)

Actually, so it won't affect the transactions table the value set in there.

mikheevm commented on src/utils/humanReadableValue.ts:
```diff
@@ -0,0 +1,5 @@
+import { BigNumber } from 'bignumber.js'
```
there is quite similar (by purpose) function in `src/logic/tokens/utils`. Should we maybe move this one there too?

fernandomg commented on src/components/Collapse/index.tsx:
```diff
@@ -7,14 +7,29 @@ import styled from 'styled-components'
 
+const HeaderWrapper = styled.div``
+
+const Title = styled.div``
```
actually, here I prefer it to be `null` as it can be a simple text or a ReactNode.

Also, if it's `null` it won't generate a new node in the DOM and I don't need to make further checks to prevent that, as in the `''` case.

mikheevm replied:
I think you were talking about `description = null` thing

fernandomg replied:
hehehe, yes :grimacing:

fernandomg commented on src/logic/contracts/methodIds.ts:
```diff
@@ -59,39 +60,28 @@ const METHOD_TO_ID = {
         parameters: [
-          { name: 'oldOwner', value: decodedParameters[1] },
-          { name: 'newOwner', value: decodedParameters[2] },
+          { name: 'oldOwner', type: '', value: decodedParameters[1] },
```
It shouldn't be optional. But I added as an empty string as it's not used in the app, but the `parameters` key always returns `name`, `type`, and `value`.

I'll add the proper types

fernandomg commented on src/routes/safe/store/models/types/transactions.ts:
```diff
@@ -0,0 +1,184 @@
+  owner: string
+  submissionDate: string
+  transactionHash: string | null
+  confirmationType: keyof typeof ConfirmationType
```
it looks like it is required: [TSPlayground](https://www.typescriptlang.org/play?#code/KYOwrgtgBAyglgcxAQwC5gE7ACoE8AOwUA3gLABQUUAwgPIBy2ASgILXYD6MAkgOL0tsAVSYBRKAF4oAcjqNW7LnwHCx0gDQUqLAAo6mtAGqiAIhwASLGOckzd+o6YtXzGrVFG0Wt6Z5ZvKD2xzJX4fUWDQ+gCAXwoKVAIiGGxbAGtgXAB7ADMoRMJc2EQUdCw8QnjyOBBUYAwc5ABjImRqLIh8ABtgAA9aECIyQIBnRAAuYqQ0TBwkzUDUAAtgeARJlIo48gomrJAR1ChkEFoAIwArYCbUSbaO7r6BoilhqjH1mT8AqmXVia+XmkWyqQA)

I've created those types using the collection returned by the tx service. What I mean, I was using the object and was evaluating its validity inline.

mikheevm replied:
enums are not designed to use this way, here is a playground with the correct usage of enum: https://www.typescriptlang.org/play/?ssl=1&ssc=1&pln=22&pc=1#code/KYOwrgtgBAyglgcxAQwC5gE7ACoE8AOwUA3gLABQUUAwgPIBy2ASgILXYD6MAkgOL0tsAVSYBRKAF4oAcjqNW7LnwHCx0gDQUqLAAo6mtAGqiAIhwASLGOckzd+o6YtXzGrVFG0Wt6Z5ZvKD2xzJX4fUWDQ+gCAXwoKVAIiGGxbAGtgXAB7ADMoRMJc2EQUdCw8QnjyOBBUYAwc5ABjImRqLIh8ABtgAA9aECIyQIBnRAAuYqQ0TBwkzUCAEzR4BEnV0tmK4AWqVAALYFX17Ao48gomrJAR1ChkEFoAIwArYCbUSbaO7r6BoikwyoYzWU025SSADo-LsoMtUMcwTMIYQANp+AC6sIORwmMj80jOVSAA

mikheevm replied:
https://www.youtube.com/watch?v=JfcLkoBirZo

fernandomg replied:
but, well... you're helping me with my point. I needed to do the `keyof typeof` to be able to properly parse the JSON object returned by the backend.

mikheevm replied:
yeah, but in this case, why do we need an enum? In my understanding the desired usage of the enum would be the following:
```
enum TransactionTypes {
  outgoing = 'out',
  incoming = 'in'
}

function makeTransaction(type: TransactionTypes)

makeTransaction('out') - errors
makeTransaction(TransactionTypes.outgoing) - works
```

I ran into some issue in safe-apps-sdk with enums (don't recall what it was) and it was fixed by creating an object with constant values. I think using object is a better way.
Also, there are some caveats of how enums are compiled, see: https://github.com/babel/babel/issues/8741

fernandomg replied:
How can this affect us? We're not exposing them like in a library and, whenever we need them, we enforce the use of the `enum`, that was my intention.

Actually, if I remove the `keyof typeof` from the equation, I shouldn't be an issue, as the use of the string value happens on runtime (with the data returned from the backend).

If you're ok, and it works fine, I'll remove the `keyof typeof` and will stay with the enums

mikheevm replied:
my point is that the usage of enums wasn't really needed there, but ok

gnosis-info commented:
Travis automatic deployment:
 https://pr1106--safereact.review.gnosisdev.com/app

gnosis-info commented:
Travis automatic deployment:
 https://pr1106--safereact.review.gnosisdev.com/app

gnosis-info commented:
Travis automatic deployment:
 https://pr1106--safereact.review.gnosisdev.com/app

gnosis-info commented:
Travis automatic deployment:
 https://pr1106--safereact.review.gnosisdev.com/app

gnosis-info commented:
Travis automatic deployment:
 https://pr1106--safereact.review.gnosisdev.com/app

gnosis-info commented:
Travis automatic deployment:
 https://pr1106--safereact.review.gnosisdev.com/app

mikheevm commented on src/logic/contracts/methodIds.ts:
```diff
@@ -59,39 +60,28 @@ const METHOD_TO_ID = {
-
 type SafeDecodedParams = {
-  [key in SafeMethods]?: DecodedValues
+  [key in SafeMethods]?: Record<string, string>
```
Why was this changed from
```
type DecodedValues = Array<{	
  name: string	
  type?: string	
  value: string	
}>	
``` 
to more loosely typed 
```
Record<string, string>
```
?

fernandomg replied:
because that was wrong.

mikheevm replied:
but to me it seems that all the decodings still follow the same structure?
```
{ name: 'newOwner', type: 'address', value: decodedParameters[2] },
```

fernandomg replied:
yes... there's a big confusion around names. The returned type is `DataDecoded` which has that structure you're pointing here. And corresponds to the existing `dataDecoded` value returned by the backend, and is also the same structure we build in the app (this is kinda legacy code, but is used by the _pending_ txs) by calling `decodeMethods`.

```typescript
export interface DataDecoded {
  method: DataDecodedMethod
  parameters: Parameter[]
}
```

On the other hand, is `decodedParams`, that uses the `DecodedParam` type created in this file (messy and confusing all this) and is the structure already existing in the Safe, that is used by that not so good loved `getTxData` function, used to classify txs.

---

So, bottom line, I'll try to move the things around a bit, so it's a bit more explicit. But typing is fine.

mikheevm commented on src/logic/contracts/methodIds.ts:
```diff
@@ -59,39 +60,28 @@ const METHOD_TO_ID = {
 
 type TokenDecodedParams = {
-  [key in TokenMethods]?: DecodedValues
+  [key in TokenMethods]?: Record<string, string>
```
same question here

mikheevm commented on src/logic/currencyValues/store/selectors/index.ts:
```diff
@@ -1,46 +1,41 @@
 
+const currencyValueSelector = <K extends keyof CurrencyRateValue>(key: K) => (
+  currencyValuesMap?: CurrencyReducerMap,
+): CurrencyRateValue[K] => currencyValuesMap?.get(key) || null
```
is the return type ok with null being returned? also, you can use notSetValue param for `.get`, i've pointed that already a lot of times

fernandomg replied:
I get what you say about the `notSetValue`, will fix the `CurrencyReducerMap`, but won't use it here.

What will happen if `currencyValuesMap` is `undefined`? With your proposed solution this will be returning `undefined`, a `value` or `null`.

I'll remove `null` from the _equation_ here.

mikheevm replied:
why could it be undefined? we should have a default value for that (= empty map)

fernandomg replied:
if you check what was before:

```typescript
export const safeFiatBalancesListSelector = createSelector(safeFiatBalancesSelector, (currencyValuesMap) => {
  if (!currencyValuesMap) return
  return currencyValuesMap.get('currencyBalances') ? currencyValuesMap.get('currencyBalances') : List([])
})

export const currentCurrencySelector = createSelector(safeFiatBalancesSelector, (currencyValuesMap) =>
  currencyValuesMap ? currencyValuesMap.get('selectedCurrency') : null,
)

export const currencyRateSelector = createSelector(safeFiatBalancesSelector, (currencyValuesMap) =>
  currencyValuesMap ? currencyValuesMap.get('currencyRate') : null,
)
```

...we had a mixed type of returns values.

I think that this solution works fine for the three selectors and makes it less verbose and easy to reason about. That was my motivation, actually.

mikheevm commented on src/routes/safe/components/Transactions/TxsTable/ExpandedTx/TxDescription/CustomDescription.tsx:
```diff
@@ -0,0 +1,183 @@
+
+  return rawTx.multiSendTx ? (
+    <Block className={classes.multiSendTxData} data-testid={TRANSACTIONS_DESC_CUSTOM_DATA_TEST_ID}>
+      {extractMultiSendDecodedData(rawTx).txDetails?.map((tx, index) => {
```
could we extract this outside of the jsx to improve readability?

fernandomg replied:
no sure I manage to do that but review it made me clean up a bit more the code around this

mikheevm commented on src/routes/safe/store/actions/transactions/utils/transferDetails.ts:
```diff
@@ -0,0 +1,84 @@
+} from './transferDetails.d'
+import { humanReadableValue } from 'src/logic/tokens/utils/humanReadableValue'
+
+const isIncomingTransfer = (transfer: Transfer): boolean => {
+  const state = store.getState()
+  const safeAddress = safeParamAddressFromStateSelector(state)
+  return sameAddress(transfer.to, safeAddress)
+}
```
not really a fan of accessing store like this, this should be done only if there's no other option. why couldn't we just pass safeAddress there?

fernandomg replied:
actually this is not being used, as `transfers` only come as part of the `allTransactions` returned data.

I added a TODO here, so we can address this in the future

gnosis-info commented:
Travis automatic deployment:
 https://pr1106--safereact.review.gnosisdev.com/app

gnosis-info commented:
Travis automatic deployment:
 https://pr1106--safereact.review.gnosisdev.com/app

gnosis-info commented:
Travis automatic deployment:
 https://pr1106--safereact.review.gnosisdev.com/app

francovenica commented:
I repeated the same test I did the last time, trying every app possible in rinkeby and checking that they look fine in the tx and its details

The safe: https://pr1106--safereact.review.gnosisdev.com/app/#/safes/0x9913B9180C20C6b0F21B6480c84422F6ebc4B808/transactions

![image.png](https://images.zenhubusercontent.com/5e14b8e3edee21d6f9aa6c8a/7fb22040-3381-49ff-9cc2-572580fff2f7)

![image.png](https://images.zenhubusercontent.com/5e14b8e3edee21d6f9aa6c8a/d89855ed-22ec-4043-b8d9-e135e250fa4b)

gnosis-info commented:
Travis automatic deployment:
 https://pr1106--safereact.review.gnosisdev.com/app

gnosis-info commented:
Travis automatic deployment:
 https://pr1106--safereact.review.gnosisdev.com/app

fernandomg commented:
@mikheevm, please, don't forget to review recent changes

mikheevm commented on src/logic/addressBook/store/reducer/types/addressBook.d.ts:
```diff
@@ -15,7 +15,7 @@ interface AddressBookReducerStateSerialized extends AddressBookReducerState {
 export interface AddressBookMap extends Map<string> {
   toJS(): AddressBookMapSerialized
-  get(key: string): List<AddressBookEntryRecord>
+  get(key: string, notSetValue: unknown): List<AddressBookEntryRecord>
```
return type is a little bit off here, it can also return notSetValue

francovenica commented:
Tried one more time in a new safe and incognito mode and it worked fine
I've tried Compound, Sablier, Idle and The Tx builder (replicating a compound contract method).

Looks good to me

Safe for testing: https://pr1106--safereact.review.gnosisdev.com/app/#/safes/0x5D0edD9272a653Bf6b3b49e9856c7030942d2003/transactions

![image.png](https://images.zenhubusercontent.com/5e14b8e3edee21d6f9aa6c8a/e28b4639-24ff-4ba1-9cf0-66b8890ba958)

gnosis-info commented:
Travis automatic deployment:
 https://pr1106--safereact.review.gnosisdev.com/app
